modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -54,32 +56,50 @@ const restoreFakeSelection = (editor: Editor) => {
   editor.selection.setRng(editor.selection.getRng());
 };
 
-const toggleReadOnly = (editor: Editor, state: boolean): void => {
+const setCommonEditorCommands = (editor: Editor, state: boolean): void => {
+  setEditorCommandState(editor, 'StyleWithCSS', state);
+  setEditorCommandState(editor, 'enableInlineTableEditing', state);
+  setEditorCommandState(editor, 'enableObjectResizing', state);
+};
+
+const setEditorReadonly = (editor: Editor) => {
+  editor.readonly = true;
+  editor.selection.controlSelection.hideResizeRect();
+  editor._selectionOverrides.hideFakeCaret();
+  removeFakeSelection(editor);
+  setCommonEditorCommands(editor, true);
+};
+
+const unsetEditorReadonly = (editor: Editor, body: SugarElement<HTMLElement>) => {
+  editor.readonly = false;
+  if (editor.hasEditableRoot()) {
+    setContentEditable(body, true);
+  }
+  switchOnContentEditableTrue(body);
+  setCommonEditorCommands(editor, false);
+  if (EditorFocus.hasEditorOrUiFocus(editor)) {
+    editor.focus();
+  }
+  restoreFakeSelection(editor);
+  editor.nodeChanged();
+};
+
+const toggleReadOnly = (editor: Editor, readOnlyMode: EditorReadOnlyType): void => {
   const body = SugarElement.fromDom(editor.getBody());
+  const shouldSetReadOnly = Type.isBoolean(readOnlyMode) ? readOnlyMode : true;
 
-  toggleClass(body, 'mce-content-readonly', state);
+  toggleClass(body, 'mce-content-readonly', shouldSetReadOnly);
 
-  if (state) {
-    editor.selection.controlSelection.hideResizeRect();
-    editor._selectionOverrides.hideFakeCaret();
-    removeFakeSelection(editor);
-    editor.readonly = true;
-    setContentEditable(body, false);
-    switchOffContentEditableTrue(body);
-  } else {
-    editor.readonly = false;
-    if (editor.hasEditableRoot()) {
-      setContentEditable(body, true);
-    }
-    switchOnContentEditableTrue(body);
-    setEditorCommandState(editor, 'StyleWithCSS', false);
-    setEditorCommandState(editor, 'enableInlineTableEditing', false);
-    setEditorCommandState(editor, 'enableObjectResizing', false);
-    if (EditorFocus.hasEditorOrUiFocus(editor)) {
-      editor.focus();
+  const shouldSetContentEditableTrue = () => Type.isBoolean(readOnlyMode) ? readOnlyMode : !Obj.get(readOnlyMode, 'selectionEnabled').getOr(true);
+
+  if (shouldSetReadOnly) {
+    setEditorReadonly(editor);
+    if (shouldSetContentEditableTrue()) {
```

Comment:
This doesn't need to be a function if you do it inside the first `if`. I also think `Obj.get` is overkill for this scenario? Just `readonlyMode.selectionEnabled` would work for the `if` condition where this is used.
```suggestion
  if (shouldSetReadOnly) {
    const shouldSetContentEditableTrue = Type.isBoolean(readOnlyMode) ? readOnlyMode : readonlyMode.selectionEnabled;
    setEditorReadonly(editor);
    if (shouldSetContentEditableTrue) {
```